Display answers on summary screen

benjefferies opened the pull request:
### What is the context of this PR?

Describe what you have changed and why, link to other PRs or Issues as appropriate.
[Trello card](https://trello.com/c/bxYkANFg/591-2-rsi-employee-questions-heading-wrong-level)
Changing the quarterly retail survey index schema employee section. The section is now configured as multiple answers to the question `On {date} what was the number of employees?`.
The summary page has changed to support showing multiple answers for a question.
The summary page model had to change to support answers on the summary page
### How to review

Describe the steps required to test the changes (include screenshots if appropriate).
- [ ] Run multiple-answers functional test `npm run test_functional -- --spec=multiple-answers`
- Convert employee questions to answers on survey on rsi 112 survey
- Modify summary.html to display answers on summary
- Add functional test

benjefferies committed:
Display answers on summary screen

* Convert employee questions to answers on survey on rsi 112 survey
* Modify summary.html to display answers on summary
* Add functional test

iwootten commented on app/templates/summary.html:
```diff
@@ -22,13 +22,27 @@ <h2 class="header__subtitle mars">Please check carefully before submission.</h2>
 
-        <dt class="summary__question summary__question--{{item.type}}" id="{{id}}">{{item.question | safe}}</dt>
+        <dt class="summary__question summary__question--{{item.type}}" id="{{id}}">
+          {% if item.state_answers|length == 1 %}
```
I feel that this kind of conditional logic should be kept away from the frontend templates and instead wrapped up in SummaryItem behaviour.

benjefferies replied:
I think you're right, @hamishtaplin also mentioned the same thing. I'm currently trying to figure out how 'answers' and 'items' fit in as 'SummaryItems'. Any suggestions are welcomed!

hamishtaplin replied:
@benjefferies @iwootten Totally agree. Also, this logic isn't right as both of these things would need to be displayed anyway.

benjefferies replied:
It's worth taking another look at the changes. I have minimised the logic in the template now although there is still a conditional check due to the dual nature of the date range question/answer

benjefferies committed:
Merge remote-tracking branch 'origin/master' into display-answers-on-summary

benjefferies committed:
Add tests for SummaryItem

Unit tests to cover scenarios
* Creating SummaryItem with one sub item
* Creating SummaryItem with date range sub item
* Creating SummaryItem with multiple date range sub items
* Creating SummaryItem with multiple sub items

benjefferies committed:
Renaming sub item question to answer_title

hamishtaplin commented on app/templates/summary.html:
```diff
@@ -30,51 +30,60 @@ <h2 class="header__subtitle mars">Please check carefully before submission.</h2>
+            {% set item_id = id ~ "-" ~ loop.index0 %}
+            {% if item.sub_items|length > 1 %}
+              <dt class="summary__question summary__answer__title--{{sub_item.type}}" id="{{item_id}}">{{sub_item.answer_title | safe}}</dt>
+              <dd class="summary__answer summary__answer--{{item.type}}">
```
@benjefferies: Just a heads up: this isn't valid html as the `dt/dd` pair needs to be directly nested inside a `dl`. Also the `dd` tag isn't closed anywhere.

This bit of template no longer makes sense semantically anyway so I'm going to have to refactor the whole thing.

hamishtaplin committed:
move summary items into their own partial

hamishtaplin committed:
added more template partials for summary items

hamishtaplin committed:
implemented improved semantics/styling for summary page with multiple answers

hamishtaplin committed:
fixed merge conflict

benjefferies committed:
Fixing question formatter unit tests

The unit tests were failing due to changes to the summary model changing
which now using the questionnaire state utility method to find the
question and answer schema. The questionnaire state was `None` causing
invocation exceptions.

The test has been refactored to use mocking to write less code to
achieve the fix

hamishtaplin committed:
Merge branch 'master' into display-answers-on-summary

ajmaddaford commented on app/data/1_0112.json:
```diff
@@ -357,44 +320,33 @@
-                                                }
-                                            }
-                                        }
+                                              "display": {},
```
The alignment of the JSON is a bit out (extra spaces)

benjefferies replied:
Good spot, fixed.

ajmaddaford commented on tests/app/templating/summary/formatters/test_summary_item.py:
```diff
@@ -0,0 +1,71 @@
+
+    def test_create_summary_item(self):
+        # Given
+        schema = mock.MagicMock()
```
I think just calling this `schema` is confusing, `test_question_formatters.py` uses `question_schema`

benjefferies replied:
Good point, fixed.

ajmaddaford commented on tests/functional/spec/multiple-answers.spec.js:
```diff
@@ -0,0 +1,38 @@
+    expect(multipleAnswerSummaryPage.getSurname()).to.equal('Solo')
+  })
+
+  it('Given I complete a survey has multiple answers for a question when I edit an answer then I appear on the page to edit my answer', function() {
```
This doesn't read very well. I think it should either be 'Given I complete a survey **that** has...` or just drop the first bit and have 'Given a survey has multiple answers...'

benjefferies replied:
Good spot, fixed

benjefferies committed:
Fix schema formatting and BDD scenario

benjefferies committed:
Renaming schema variable to question_schema

We currently use schema as a variable name generically but it would add
clarity by naming schema variables as a more specific version i.e.
question_schema or answer_schema_item

hamishtaplin committed:
improved partials for summary screen by providing a default

hamishtaplin committed:
Merge branch 'display-answers-on-summary' of github.com:ONSdigital/eq-survey-runner into display-answers-on-summary

ajmaddaford reviewed (approved):
LGTM 👍
